### PR TITLE
Fix date format returned by REST API

### PR DESCRIPTION
**Pull request by mprimi**

### CODE REVIEW

File: genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java [hunk 1]
```diff
@@ -34,6 +41,11 @@
  */
 public class GenieDateFormat extends StdDateFormat {
 
+    /**
+     * Pattern to validate (e.g., in tests) the expected date format this class produces.
+     */
+    @VisibleForTesting
+    public static final Pattern VALID_PATTERN = Pattern.compile("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}.\\d{3}Z$");
```
Comment (tgianos) [inline genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java]: does it need to be public or could you just make it package protected?

File: genie-common/src/test/java/com/netflix/genie/common/util/GenieDateFormatUnitTests.java [hunk 2]
```diff
@@ -36,16 +42,81 @@
 @Category(UnitTest.class)
 public class GenieDateFormatUnitTests {
 
+    private static final String EXPECTED_DATE_STRING = "1970-01-01T00:00:00.000Z";
+    private static final Date DATE = new Date(0);
+    // Limit random dates used in test between 'epoch' and 'epoch + 1000 years'
+    private static final long MAX_RANDOM_DATE = TimeUnit.MILLISECONDS.convert(365 * 1000, TimeUnit.DAYS);
```
Comment (tgianos) [inline genie-common/src/test/java/com/netflix/genie/common/util/GenieDateFormatUnitTests.java]: You might want to add a supplier to `RandomSuppliers` i have for tests.

File: genie-web/src/test/java/com/netflix/genie/web/controllers/RestControllerIntegrationTestsBase.java [hunk 3]
```diff
@@ -115,6 +117,21 @@
         .setDateFormat(new GenieDateFormat())
         .registerModule(new Jdk8Module());
 
+    protected final Matcher<String> dateFormatMatcher = new BaseMatcher<String>() {
```
Comment (tgianos) [inline genie-web/src/test/java/com/netflix/genie/web/controllers/RestControllerIntegrationTestsBase.java]: This part should probably be ported into `master` after this is merged

File: genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java [hunk 4]
```diff
@@ -65,4 +77,27 @@ public StringBuffer format(final Date date, final StringBuffer toAppendTo, final
             )
         );
     }
+
+    /**
+     * JsonSerializer to output (Optional) Date using GenieDateFormat.
+     *
+     * @since 3.3.11
+     */
+    public static class OptionalDateJsonSerializer extends JsonSerializer<Optional<Date>> {
```
Comment (tgianos) [inline genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java]: I don't see you annotating any `Optional` dates as you have the annotation on members themselves. Does jackson call the getters instead and that's why you have it like this?

File: genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java (hunk 1 above)
Comment (mprimi) [reply genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java]: It's used from the `web` module/package.

File: genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java (hunk 4 above)
Comment (mprimi) [reply genie-common/src/main/java/com/netflix/genie/common/util/GenieDateFormat.java]: Yes

File: genie-web/src/test/java/com/netflix/genie/web/controllers/RestControllerIntegrationTestsBase.java (hunk 3 above)
Comment (mprimi) [reply genie-web/src/test/java/com/netflix/genie/web/controllers/RestControllerIntegrationTestsBase.java]: Yes, i left the controller tests into a separate commit so we can cherry-pick

**Pull request merged**